ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -708,6 +694,50 @@ FindInfo findEntriesSince(FindToken token, long maxTotalSizeOfEntries) throws St
     }
   }
 
+  /**
+   * Validate the {@link StoreFindToken} and reset if required
+   * @param storeFindTokenAtomicReference the atomic reference to the {@link StoreFindToken}
+   * @return {@code true} is {@link StoreFindToken} was reset, {@code false} otherwise
+   */
+  private boolean resetTokenIfRequired(AtomicReference<StoreFindToken> storeFindTokenAtomicReference) {
+    boolean tokenReset = false;
+    StoreFindToken storeToken = storeFindTokenAtomicReference.get();
+    UUID remoteIncarnationId = storeToken.getIncarnationId();
+    // if incarnationId is null, for backwards compatability purposes, the token is considered as good.
+    /// if not null, we check for a match
+    if (!storeToken.getType().equals(StoreFindToken.Type.Uninitialized) &&
+        remoteIncarnationId != null && !remoteIncarnationId.equals(incarnationId)) {
+      // incarnationId mismatch, hence resetting the token to beginning
+      logger.info("Index : {} resetting offset after incarnation, new incarnation Id {}, "
+          + "incarnationId from store token {}", dataDir, incarnationId, remoteIncarnationId);
+      storeFindTokenAtomicReference.set(new StoreFindToken());
+      tokenReset = true;
+    }
+    // validate token
+    else if (storeToken.getSessionId() == null || storeToken.getSessionId().compareTo(sessionId) != 0) {
+      // the session has changed. check if we had an unclean shutdown on startup
+      if (!cleanShutdown) {
+        // if we had an unclean shutdown and the token offset is larger than the logEndOffsetOnStartup
+        // we reset the token to logEndOffsetOnStartup
+        if (!storeToken.getType().equals(StoreFindToken.Type.Uninitialized)
+            && storeToken.getOffset().compareTo(logEndOffsetOnStartup) > 0) {
+          logger.info("Index : " + dataDir + " resetting offset after not clean shutdown " + logEndOffsetOnStartup
+              + " before offset " + storeToken.getOffset());
+          storeFindTokenAtomicReference.set(new StoreFindToken(logEndOffsetOnStartup, sessionId, incarnationId));
+          tokenReset = true;
+        }
+      } else if (!storeToken.getType().equals(StoreFindToken.Type.Uninitialized)
+          && storeToken.getOffset().compareTo(logEndOffsetOnStartup) > 0) {
+        logger.error(
+            "Index : " + dataDir + " invalid token. Provided offset is outside the log range after clean shutdown");
+        // if the shutdown was clean, the offset should always be lesser or equal to the logEndOffsetOnStartup
+        throw new IllegalArgumentException(
+            "Invalid token. Provided offset is outside the log range after clean shutdown");
+      }
+    }
+    return tokenReset;
+  }
```

Comment:
having this method return the processed token as @vgkholla suggested should make life easier. The token that is reset should already include the information about inclusiveness.